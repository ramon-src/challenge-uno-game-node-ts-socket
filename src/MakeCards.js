import {
  MAX_CHANGE_4_CARDS,
  MAX_NUMBER_NORMAL_CARDS,
  COLORS,
  REVERT,
  BLOCKED,
  MAX_CHANGE_COLOR_CARDS,
  CHANGE_COLOR,
  ADD_2_CARD,
  ADD_4_CARD
} from "./constants";

function MakeCards() {
  this.cards = [];

  this.build = function () {
    for (let number = 0; number < MAX_NUMBER_NORMAL_CARDS + 1; number++) {
      Object.values(COLORS).forEach((color) => {
        this.cards.push({ number, color });
        this.cards.push({ number, color });
      });
    }

    Object.values(COLORS).forEach((color) => {
      this.cards.push({ color, ...ADD_2_CARD });
      this.cards.push({ color, ...ADD_2_CARD });

      this.cards.push({ color, ...REVERT });
      this.cards.push({ color, ...REVERT });

      this.cards.push({ color, ...BLOCKED });
      this.cards.push({ color, ...BLOCKED });
    });

    for (let number1 = 0; number1 < MAX_CHANGE_COLOR_CARDS; number1++) {
      this.cards.push({ ...CHANGE_COLOR });
    }

    for (let number2 = 0; number2 < MAX_CHANGE_4_CARDS; number2++) {
      this.cards.push({ ...ADD_4_CARD });
    }
    return this.cards;
  };
}

export default MakeCards;
