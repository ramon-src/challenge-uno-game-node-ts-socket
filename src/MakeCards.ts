import { Card } from "./Interfaces";
import {
  MAX_CHANGE_4_CARDS,
  MAX_NUMBER_NORMAL_CARDS,
  COLORS,
  REVERT,
  BLOCKED,
  MAX_CHANGE_COLOR_CARDS,
  CHANGE_COLOR,
  ADD_2_CARD,
  ADD_4_CARD
} from "./constants";

function MakeCards() {
  this.cards = [] as Card[];

  this.build = function (): Card[] {
    enum Type {
      NUMBER = "NUMBER",
      EFFECT = "EFFECT"
    }

    const makeIdCardWithColors = ({ typeId, order, value, color }): string => {
      return `CARD_${Type[typeId]}_${value}_COLOR_${color}_${order}`;
    };

    for (let number = 0; number < MAX_NUMBER_NORMAL_CARDS + 1; number++) {
      Object.values(COLORS).forEach((color) => {
        this.cards.push({
          id: makeIdCardWithColors({
            typeId: Type.NUMBER,
            order: 1,
            value: number,
            color
          }),
          number,
          color
        });
        this.cards.push({
          id: makeIdCardWithColors({
            typeId: Type.NUMBER,
            order: 2,
            value: number,
            color
          }),
          number,
          color
        });
      });
    }

    Object.values(COLORS).forEach((color) => {
      this.cards.push({
        id: makeIdCardWithColors({
          typeId: Type.EFFECT,
          order: 1,
          value: "ADD_2",
          color
        }),
        color,
        ...ADD_2_CARD
      });
      this.cards.push({
        id: makeIdCardWithColors({
          typeId: Type.EFFECT,
          order: 2,
          value: "ADD_2",
          color
        }),
        color,
        ...ADD_2_CARD
      });

      this.cards.push({
        id: makeIdCardWithColors({
          typeId: Type.EFFECT,
          order: 1,
          value: "REVERT",
          color
        }),
        color,
        ...REVERT
      });
      this.cards.push({
        id: makeIdCardWithColors({
          typeId: Type.EFFECT,
          order: 2,
          value: "REVERT",
          color
        }),
        color,
        ...REVERT
      });

      this.cards.push({
        id: makeIdCardWithColors({
          typeId: Type.EFFECT,
          order: 1,
          value: "BLOCKED",
          color
        }),
        color,
        ...BLOCKED
      });
      this.cards.push({
        id: makeIdCardWithColors({
          typeId: Type.EFFECT,
          order: 2,
          value: "BLOCKED",
          color
        }),
        color,
        ...BLOCKED
      });
    });

    const makeIdWithoutColor = (orderNumber, effectName): string =>
      `CARD_EFFECT_${effectName}_${orderNumber}`;

    for (let number1 = 0; number1 < MAX_CHANGE_COLOR_CARDS; number1++) {
      this.cards.push({
        id: makeIdWithoutColor(number1, "CHANGE_COLOR"),
        ...CHANGE_COLOR
      });
    }

    for (let number2 = 0; number2 < MAX_CHANGE_4_CARDS; number2++) {
      this.cards.push({
        id: makeIdWithoutColor(number2, "ADD_4"),
        ...ADD_4_CARD
      });
    }
    return this.cards;
  };
}

export default MakeCards;
