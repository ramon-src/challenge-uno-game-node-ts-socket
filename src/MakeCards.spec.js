import MakeCards from "./MakeCards";

describe("MakeCards", () => {
  it("should make all the cards", () => {
    const cards = new MakeCards().build();
    expect(cards).toMatchSnapshot();
  });
});
