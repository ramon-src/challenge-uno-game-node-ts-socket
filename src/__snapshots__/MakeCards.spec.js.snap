// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MakeCards should make all the cards 1`] = `
Array [
  Object {
    "color": "RED",
    "number": 0,
  },
  Object {
    "color": "RED",
    "number": 0,
  },
  Object {
    "color": "BLUE",
    "number": 0,
  },
  Object {
    "color": "BLUE",
    "number": 0,
  },
  Object {
    "color": "GREEN",
    "number": 0,
  },
  Object {
    "color": "GREEN",
    "number": 0,
  },
  Object {
    "color": "YELLOW",
    "number": 0,
  },
  Object {
    "color": "YELLOW",
    "number": 0,
  },
  Object {
    "color": "RED",
    "number": 1,
  },
  Object {
    "color": "RED",
    "number": 1,
  },
  Object {
    "color": "BLUE",
    "number": 1,
  },
  Object {
    "color": "BLUE",
    "number": 1,
  },
  Object {
    "color": "GREEN",
    "number": 1,
  },
  Object {
    "color": "GREEN",
    "number": 1,
  },
  Object {
    "color": "YELLOW",
    "number": 1,
  },
  Object {
    "color": "YELLOW",
    "number": 1,
  },
  Object {
    "color": "RED",
    "number": 2,
  },
  Object {
    "color": "RED",
    "number": 2,
  },
  Object {
    "color": "BLUE",
    "number": 2,
  },
  Object {
    "color": "BLUE",
    "number": 2,
  },
  Object {
    "color": "GREEN",
    "number": 2,
  },
  Object {
    "color": "GREEN",
    "number": 2,
  },
  Object {
    "color": "YELLOW",
    "number": 2,
  },
  Object {
    "color": "YELLOW",
    "number": 2,
  },
  Object {
    "color": "RED",
    "number": 3,
  },
  Object {
    "color": "RED",
    "number": 3,
  },
  Object {
    "color": "BLUE",
    "number": 3,
  },
  Object {
    "color": "BLUE",
    "number": 3,
  },
  Object {
    "color": "GREEN",
    "number": 3,
  },
  Object {
    "color": "GREEN",
    "number": 3,
  },
  Object {
    "color": "YELLOW",
    "number": 3,
  },
  Object {
    "color": "YELLOW",
    "number": 3,
  },
  Object {
    "color": "RED",
    "number": 4,
  },
  Object {
    "color": "RED",
    "number": 4,
  },
  Object {
    "color": "BLUE",
    "number": 4,
  },
  Object {
    "color": "BLUE",
    "number": 4,
  },
  Object {
    "color": "GREEN",
    "number": 4,
  },
  Object {
    "color": "GREEN",
    "number": 4,
  },
  Object {
    "color": "YELLOW",
    "number": 4,
  },
  Object {
    "color": "YELLOW",
    "number": 4,
  },
  Object {
    "color": "RED",
    "number": 5,
  },
  Object {
    "color": "RED",
    "number": 5,
  },
  Object {
    "color": "BLUE",
    "number": 5,
  },
  Object {
    "color": "BLUE",
    "number": 5,
  },
  Object {
    "color": "GREEN",
    "number": 5,
  },
  Object {
    "color": "GREEN",
    "number": 5,
  },
  Object {
    "color": "YELLOW",
    "number": 5,
  },
  Object {
    "color": "YELLOW",
    "number": 5,
  },
  Object {
    "color": "RED",
    "number": 6,
  },
  Object {
    "color": "RED",
    "number": 6,
  },
  Object {
    "color": "BLUE",
    "number": 6,
  },
  Object {
    "color": "BLUE",
    "number": 6,
  },
  Object {
    "color": "GREEN",
    "number": 6,
  },
  Object {
    "color": "GREEN",
    "number": 6,
  },
  Object {
    "color": "YELLOW",
    "number": 6,
  },
  Object {
    "color": "YELLOW",
    "number": 6,
  },
  Object {
    "color": "RED",
    "number": 7,
  },
  Object {
    "color": "RED",
    "number": 7,
  },
  Object {
    "color": "BLUE",
    "number": 7,
  },
  Object {
    "color": "BLUE",
    "number": 7,
  },
  Object {
    "color": "GREEN",
    "number": 7,
  },
  Object {
    "color": "GREEN",
    "number": 7,
  },
  Object {
    "color": "YELLOW",
    "number": 7,
  },
  Object {
    "color": "YELLOW",
    "number": 7,
  },
  Object {
    "color": "RED",
    "number": 8,
  },
  Object {
    "color": "RED",
    "number": 8,
  },
  Object {
    "color": "BLUE",
    "number": 8,
  },
  Object {
    "color": "BLUE",
    "number": 8,
  },
  Object {
    "color": "GREEN",
    "number": 8,
  },
  Object {
    "color": "GREEN",
    "number": 8,
  },
  Object {
    "color": "YELLOW",
    "number": 8,
  },
  Object {
    "color": "YELLOW",
    "number": 8,
  },
  Object {
    "color": "RED",
    "number": 9,
  },
  Object {
    "color": "RED",
    "number": 9,
  },
  Object {
    "color": "BLUE",
    "number": 9,
  },
  Object {
    "color": "BLUE",
    "number": 9,
  },
  Object {
    "color": "GREEN",
    "number": 9,
  },
  Object {
    "color": "GREEN",
    "number": 9,
  },
  Object {
    "color": "YELLOW",
    "number": 9,
  },
  Object {
    "color": "YELLOW",
    "number": 9,
  },
  Object {
    "color": "RED",
    "effect": [Function],
  },
  Object {
    "color": "RED",
    "effect": [Function],
  },
  Object {
    "color": "RED",
    "effect": [Function],
  },
  Object {
    "color": "RED",
    "effect": [Function],
  },
  Object {
    "color": "RED",
    "effect": [Function],
  },
  Object {
    "color": "RED",
    "effect": [Function],
  },
  Object {
    "color": "BLUE",
    "effect": [Function],
  },
  Object {
    "color": "BLUE",
    "effect": [Function],
  },
  Object {
    "color": "BLUE",
    "effect": [Function],
  },
  Object {
    "color": "BLUE",
    "effect": [Function],
  },
  Object {
    "color": "BLUE",
    "effect": [Function],
  },
  Object {
    "color": "BLUE",
    "effect": [Function],
  },
  Object {
    "color": "GREEN",
    "effect": [Function],
  },
  Object {
    "color": "GREEN",
    "effect": [Function],
  },
  Object {
    "color": "GREEN",
    "effect": [Function],
  },
  Object {
    "color": "GREEN",
    "effect": [Function],
  },
  Object {
    "color": "GREEN",
    "effect": [Function],
  },
  Object {
    "color": "GREEN",
    "effect": [Function],
  },
  Object {
    "color": "YELLOW",
    "effect": [Function],
  },
  Object {
    "color": "YELLOW",
    "effect": [Function],
  },
  Object {
    "color": "YELLOW",
    "effect": [Function],
  },
  Object {
    "color": "YELLOW",
    "effect": [Function],
  },
  Object {
    "color": "YELLOW",
    "effect": [Function],
  },
  Object {
    "color": "YELLOW",
    "effect": [Function],
  },
  Object {
    "effect": [Function],
  },
  Object {
    "effect": [Function],
  },
  Object {
    "effect": [Function],
  },
  Object {
    "effect": [Function],
  },
  Object {
    "effect": [Function],
  },
  Object {
    "effect": [Function],
  },
  Object {
    "effect": [Function],
  },
  Object {
    "effect": [Function],
  },
]
`;
